app/ore/project/io/PluginFile.scala

```diff
@@ -78,58 +75,64 @@ class PluginFile(private var _path: Path, val signaturePath: Path, val user: Use
     *
     * TODO: More validation on PluginMetadata results (null checks, etc)
     *
-    * @return Result of parse
+    * @return Plugin metadata or an error message
     */
   @throws[InvalidPluginFileException]
-  def loadMeta()(implicit messages: Messages): PluginMetadata = {
+  def loadMeta()(implicit messages: Messages): Either[PluginFileData, String] = {
+    val fileNames = PluginFileData.getFileNames
+
     var jarIn: JarInputStream = null
     try {
       // Find plugin JAR
       jarIn = new JarInputStream(newJarStream)
 
+      var data = List[DataValue[_]]()
+
       // Find plugin meta file
       var entry: JarEntry = null
-      var metaFound: Boolean = false
-      breakable {
-        while ({ entry = jarIn.getNextJarEntry; entry } != null) {
-          if (entry.getName.equals(MetaFileName)) {
-            metaFound = true
-            break
-          }
+      while ( {
+        entry = jarIn.getNextJarEntry
+        entry
+      } != null) {
+        if (fileNames.contains(entry.getName)) {
+          data = data ::: PluginFileData.getData(entry.getName, new BufferedReader(new InputStreamReader(jarIn)))
```

Comment:
::: on a list in a loop is bad as it needs to go over the entire list. Can we use another data structure, or refactor this to be able to use ::?